### ISSUE TITLE
New compilation warning in core

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16620]
Migrated from [rt.perl.org#133360](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133360) (status was 'resolved')

Searchable as RT133360$

**Comment by p5pRT** [comment]
### From @dur-randir

#### Created by @dur-randir



The following warning is new in blead for the perl interpreter built
with \-DDEBUGGING \-Dusethreads&#8203;:

cc \-c \-DPERL\_CORE \-fno\-common \-DPERL\_DARWIN \-mmacosx\-version\-min=10\.9
\-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector
\-I/usr/local/include \-DPERL\_USE\_SAFE\_PUTENV \-std=c89 \-O3 \-g \-Wall
\-Werror=declaration\-after\-statement \-Werror=pointer\-arith \-Wextra
\-Wc\+\+\-compat \-Wwrite\-strings util\.c
util\.c&#8203;:2073&#8203;:1&#8203;: warning&#8203;: unused function 'S\_env\_alloc' \[\-Wunused\-function\]
S\_env\_alloc\(void \*current\, Size\_t l1\, Size\_t l2\, Size\_t l3\, Size\_t size\)
^
1 warning generated\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.29.0:

Configured by root at Mon Jun 25 01:11:54 MSK 2018.

Summary of my perl5 (revision 5 version 29 subversion 0) configuration:
  Commit id: d361a1e6288b3e67fba6cb9f5cdac3737a1e3795
  Platform:
    osname=linux
    osvers=4.9.0-6-amd64
    archname=x86_64-linux
    uname='linux dorothy 4.9.0-6-amd64 #1 smp debian 4.9.88-1+deb9u1
(2018-05-07) x86_64 gnulinux '
    config_args='-des -Dusedevel -Dcc=afl-clang-fast -Doptimize=-O3 -g
-fno-omit-frame-pointer'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='afl-clang-fast'
    ccflags ='-fno-strict-aliasing -pipe -fstack-protector-strong
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64
-D_FORTIFY_SOURCE=2'
    optimize='-O3 -g -fno-omit-frame-pointer'
    cppflags='-fno-strict-aliasing -pipe -fstack-protector-strong
-I/usr/local/include'
    ccversion=''
    gccversion='4.2.1 Compatible Clang 3.9.1 (tags/RELEASE_391/rc2)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='afl-clang-fast'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/llvm-3.9/bin/../lib/clang/3.9.1/lib
/usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu
/lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.24.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.24'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O3 -g -fno-omit-frame-pointer -L/usr/local/lib
-fstack-protector-strong'



@INC for perl 5.29.0:
    lib
    /usr/local/lib/perl5/site_perl/5.29.0/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.29.0
    /usr/local/lib/perl5/5.29.0/x86_64-linux
    /usr/local/lib/perl5/5.29.0


Environment for perl 5.29.0:
    HOME=/home/afl
    LANG=en_US.UTF-8
    LANGUAGE=en_US:en
    LC_CTYPE=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.20.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERLBREW_BASHRC_VERSION=0.78
    PERLBREW_HOME=/home/afl/.perlbrew
    PERLBREW_MANPATH=/home/afl/perlbrew/perls/perl-5.20.2/man
    PERLBREW_PATH=/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.20.2/bin
    PERLBREW_PERL=perl-5.20.2
    PERLBREW_ROOT=/home/afl/perlbrew
    PERLBREW_VERSION=0.78
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed\, 11 Jul 2018 13&#8203;:02&#8203;:43 GMT\, randir wrote&#8203;:
> This is a bug report for perl from sergey\.aleynikov@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.41 running under perl 5\.29\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> The following warning is new in blead for the perl interpreter built
> with \-DDEBUGGING \-Dusethreads&#8203;:
> 
> cc \-c \-DPERL\_CORE \-fno\-common \-DPERL\_DARWIN \-mmacosx\-version\-min=10\.9
> \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector
> \-I/usr/local/include \-DPERL\_USE\_SAFE\_PUTENV \-std=c89 \-O3 \-g \-Wall
> \-Werror=declaration\-after\-statement \-Werror=pointer\-arith \-Wextra
> \-Wc\+\+\-compat \-Wwrite\-strings util\.c
> util\.c&#8203;:2073&#8203;:1&#8203;: warning&#8203;: unused function 'S\_env\_alloc' \[\-Wunused\-
> function\]
> S\_env\_alloc\(void \*current\, Size\_t l1\, Size\_t l2\, Size\_t l3\, Size\_t
> size\)
> ^
> 1 warning generated\.

I tried a similar configuration and build on FreeBSD\-11\.1 but did \*not\* get that warning&#8203;:

\#\#\#\#\#
$ uname \-mrs
FreeBSD 11\.1\-STABLE amd64

$ clang \-\-version
FreeBSD clang version 6\.0\.0 \(tags/RELEASE\_600/final 326565\) \(based on LLVM 6\.0\.0\)
Target&#8203;: x86\_64\-unknown\-freebsd11\.1
Thread model&#8203;: posix
InstalledDir&#8203;: /usr/bin

$ \./perl \-Ilib \-V | grep config\_args
&nbsp;    config\_args='\-des \-Dusedevel \-Duseithreads \-Doptimize=\-O2 \-pipe \-fstack\-protector \-fno\-strict\-aliasing \-DDEBUGGING'

$ grep \-n 'unused function' 133360\-make\-test\-prep\-log\.txt 
$ \[no output\]
\#\#\#\#\#

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @dur-randir

On Tue\, 17 Jul 2018 12&#8203;:39&#8203;:10 \-0700\, jkeenan wrote&#8203;:
> I tried a similar configuration and build on FreeBSD\-11\.1 but did
> \*not\* get that warning&#8203;:

> $ uname \-mrs
> FreeBSD 11\.1\-STABLE amd64
> $ clang \-\-version
> FreeBSD clang version 6\.0\.0 \(tags/RELEASE\_600/final 326565\) \(based on
> LLVM 6\.0\.0\)

It's a platform\-specific \(Darwin\) issue\, not compiler\-specific\. I've investigated a bit further and have found offending commit to be 

commit 34716e2a6ee2af96078d62b065b7785c001194be
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Fri Jun 29 13&#8203;:37&#8203;:03 2018 \+0100

&nbsp;    Perl\_my\_setenv\(\); handle integer wrap

&nbsp;    RT \#133204

which intorduced S\_env\_alloc funnction\, but in a \#if defined\(PERL\_DARWIN\) branch on line 2169 still directly calls setenv\(\)\, making S\_env\_alloc unused on some platforms \(as I haven't analyzed which other platforms from that line will also get this warning\)\.

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Jul 22\, 2018 at 08&#8203;:33&#8203;:40AM \-0700\, Sergey Aleynikov via RT wrote&#8203;:
> On Tue\, 17 Jul 2018 12&#8203;:39&#8203;:10 \-0700\, jkeenan wrote&#8203;:
> > I tried a similar configuration and build on FreeBSD\-11\.1 but did
> > \*not\* get that warning&#8203;:
> 
> > $ uname \-mrs
> > FreeBSD 11\.1\-STABLE amd64
> > $ clang \-\-version
> > FreeBSD clang version 6\.0\.0 \(tags/RELEASE\_600/final 326565\) \(based on
> > LLVM 6\.0\.0\)
> 
> It's a platform\-specific \(Darwin\) issue\, not compiler\-specific\. I've investigated a bit further and have found offending commit to be 
> 
> commit 34716e2a6ee2af96078d62b065b7785c001194be
> Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> Date&#8203;:   Fri Jun 29 13&#8203;:37&#8203;:03 2018 \+0100
> 
>     Perl\_my\_setenv\(\); handle integer wrap
> 
>     RT \#133204
> 
> which intorduced S\_env\_alloc funnction\, but in a \#if defined\(PERL\_DARWIN\) branch on line 2169 still directly calls setenv\(\)\, making S\_env\_alloc unused on some platforms \(as I haven't analyzed which other platforms from that line will also get this warning\)\.

Now fixed \(hopefully\) by&#8203;:

commit 3d50648cc4f452d51a0564164a214e51dd6aad95
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Mon Jul 23 09&#8203;:15&#8203;:38 2018 \+0100

&nbsp;    Don't compile S\_env\_alloc\(\) when not used
&nbsp;    
&nbsp;    RT \#133360
&nbsp;    
&nbsp;    I recently added S\_env\_alloc\(\) as a helper function for
&nbsp;    Perl\_my\_setenv\(\); but it's not needed on all platforms\, so \#ifdef it to
&nbsp;    only be compiled on platforms where its needed\.
&nbsp;    
&nbsp;    Avoids a compiler warning\.


\-\- 
Never do today what you can put off till tomorrow\.

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Jul 22\, 2018 at 08&#8203;:33&#8203;:40AM \-0700\, Sergey Aleynikov via RT wrote&#8203;:
> On Tue\, 17 Jul 2018 12&#8203;:39&#8203;:10 \-0700\, jkeenan wrote&#8203;:
> > I tried a similar configuration and build on FreeBSD\-11\.1 but did
> > \*not\* get that warning&#8203;:
> 
> > $ uname \-mrs
> > FreeBSD 11\.1\-STABLE amd64
> > $ clang \-\-version
> > FreeBSD clang version 6\.0\.0 \(tags/RELEASE\_600/final 326565\) \(based on
> > LLVM 6\.0\.0\)
> 
> It's a platform\-specific \(Darwin\) issue\, not compiler\-specific\. I've investigated a bit further and have found offending commit to be 
> 
> commit 34716e2a6ee2af96078d62b065b7785c001194be
> Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> Date&#8203;:   Fri Jun 29 13&#8203;:37&#8203;:03 2018 \+0100
> 
>     Perl\_my\_setenv\(\); handle integer wrap
> 
>     RT \#133204
> 
> which intorduced S\_env\_alloc funnction\, but in a \#if defined\(PERL\_DARWIN\) branch on line 2169 still directly calls setenv\(\)\, making S\_env\_alloc unused on some platforms \(as I haven't analyzed which other platforms from that line will also get this warning\)\.

Now fixed \(hopefully\) by&#8203;:

commit 3d50648cc4f452d51a0564164a214e51dd6aad95
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Mon Jul 23 09&#8203;:15&#8203;:38 2018 \+0100

&nbsp;    Don't compile S\_env\_alloc\(\) when not used
&nbsp;    
&nbsp;    RT \#133360
&nbsp;    
&nbsp;    I recently added S\_env\_alloc\(\) as a helper function for
&nbsp;    Perl\_my\_setenv\(\); but it's not needed on all platforms\, so \#ifdef it to
&nbsp;    only be compiled on platforms where its needed\.
&nbsp;    
&nbsp;    Avoids a compiler warning\.


\-\- 
Never do today what you can put off till tomorrow\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
